### PR TITLE
Commit app_context project.

### PR DESCRIPTION
**Pull request by CoitThomas**

### CODE REVIEW

File: lib/remove_context.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    if iab_id
+      valid_entries << domain
+    else
+      invalid_entries << domain
+    end
+  end
+  return valid_entries, invalid_entries
+end
+
+def print_valid_entries(valid_entries, logger)
```
Comment (ProntoPolice) [inline lib/remove_context.rb]: Similar code found in remove_context.rb:25, remove_context.rb:32 (mass = 36)

File: lib/remove_context.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  valid_entries.each do |domain|
+    logger.info("#{domain} removed from database successfully.")
+  end
+  logger.info("Number of mobile app domains removed: #{valid_entries.length}")
+end
+
+def print_invalid_entries(invalid_entries, logger)
```
Comment (ProntoPolice) [inline lib/remove_context.rb]: Similar code found in remove_context.rb:25, remove_context.rb:32 (mass = 36)

File: lib/app_context.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    domains.each do |domain|
+      verified_entry = @redis_connection.get(domain)
+      verified_entries[domain] = verified_entry if verified_entry
+    end
+    verified_entries
+  end
+
+  def load(app_context_expire_time, mobile_apps)
+    mobile_apps.each_pair do |domain, sifi_category_name|
+      if @iab_categories[sifi_category_name]
```
Comment (ProntoPolice) [inline lib/app_context.rb]: Calls @iab_categories[sifi_category_name] 2 times (DuplicateMethodCall)

File: lib/app_context.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
+# Loads as many clean domains and their corresponding iab
+# category id numbers from the top 250 Apps list as possible.
+class AppContext
+  def initialize(redis_connection, iab_categories)
+    @redis_connection = redis_connection
+    @iab_categories = iab_categories
+  end
+
+  def verify_entries(mobile_apps)
```
Comment (ProntoPolice) [inline lib/app_context.rb]: Has approx 6 statements (TooManyStatements)

File: lib/app_context_config.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
+require 'yaml'
+require 'redis'
+require 'sequel'
+
+# Sets the appropriate configurations for the redis database connection, sql
+# database connection, and csv file.
+class AppContextConfig
```
Comment (ProntoPolice) [inline lib/app_context_config.rb]: Has at least 7 instance variables (TooManyInstanceVariables)

File: lib/app_context_config.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+              :redis_host,
+              :redis_port,
+              :sql_connect_string
+
+  def initialize(config_filename)
+    @configfilename = config_filename
+    set_from_file(config_filename)
+  end
+
+  def set_from_file(filename)
```
Comment (ProntoPolice) [inline lib/app_context_config.rb]: Has approx 8 statements (TooManyStatements)

File: lib/iab_categories.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  def initialize(postgres_connection)
+    @dictionary = build_iab_categories_hash(postgres_connection)
+  end
+
+  extend Forwardable
+  def_delegators :@dictionary, :[], :keys, :each
+
+  private
+
+  def build_iab_categories_hash(postgres_connection)
```
Comment (ProntoPolice) [inline lib/iab_categories.rb]: Has approx 6 statements (TooManyStatements)

Comment (ProntoPolice) [inline lib/iab_categories.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: lib/load_context.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
+$LOAD_PATH.unshift(File.expand_path('../', __FILE__))
+require 'app_context'
+require 'mobile_app'
+require 'app_context_config'
+require 'iab_categories'
+require 'logger'
+require 'redis'
+require 'parse_mobile_apps'
+
+def print_redis_entries(redis_entries, logger)
```
Comment (ProntoPolice) [inline lib/load_context.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: lib/parse_mobile_apps.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  # Used to validate the the data from each row in the file.
+  class InvalidMobileAppDataException < StandardError
+    def initialize(filename, row)
+      msg = "File #{filename} has invalid mobile app data on row '#{row}'. " \
+        "Expected keys ##{EXPECT_KEYS.join(',')}."
+      super(msg)
+    end
+  end
+
+  def self.from_csv_file(csv_filename)
```
Comment (ProntoPolice) [inline lib/parse_mobile_apps.rb]: Has approx 8 statements (TooManyStatements)

File: lib/parse_mobile_apps.rb [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+      if mobile_app_domain && sifi_category_name && !exception_note
+        mobile_apps.push(MobileApp.new(mobile_app_domain.to_s.downcase,
+                                       sifi_category_name.to_s.downcase))
+      end
+    end
+    mobile_apps
+  end
+
+  def self.valid_keys?(row)
+    EXPECT_KEYS.all? { |k| row.key?(k) }
```
Comment (ProntoPolice) [inline lib/parse_mobile_apps.rb]: Has the variable name 'k' (UncommunicativeVariableName)

File: lib/remove_context.rb [hunk 11]
```diff
@@ -0,0 +1,58 @@
+$LOAD_PATH.unshift(File.expand_path('../', __FILE__))
+require 'app_context'
+require 'io/console'
+require 'logger'
+require 'app_context_config'
+require 'iab_categories'
+require 'redis'
+
+def filter_file(redis_connection, filename)
```
Comment (ProntoPolice) [inline lib/remove_context.rb]: Has approx 8 statements (TooManyStatements)

Comment (ProntoPolice) [inline lib/remove_context.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: lib/remove_context.rb (hunk 2 above)
Comment (ProntoPolice) [inline lib/remove_context.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: lib/remove_context.rb (hunk 1 above)
Comment (ProntoPolice) [inline lib/remove_context.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: spec/spec_helper.rb [hunk 12]
```diff
@@ -0,0 +1,27 @@
+$LOAD_PATH.unshift(File.expand_path('../../tools', __FILE__))
+require 'simplecov'
+SimpleCov.start do
+  add_filter '/spec/'
+  track_files 'tools/*.rb'
+end
+
+module AppContextSpec
```
Comment (ProntoPolice) [inline spec/spec_helper.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/app_context.rb (hunk 3 above)
Comment (ProntoPolice) [inline lib/app_context.rb]: Use `next` to skip iteration.

File: lib/app_context.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+                                @iab_categories[sifi_category_name].to_s)
+      end
+    end
+  end
+
+  def remove(mobile_app_domains_array)
+    @redis_connection.del(mobile_app_domains_array)
+  end
+
+  def set_some_stuff(stuff)
```
Comment (ProntoPolice) [inline lib/app_context.rb]: Do not prefix writer method names with `set_`.

File: lib/app_context_config.rb (hunk 6 above)
Comment (ProntoPolice) [inline lib/app_context_config.rb]: Method has too many lines. [11/10]

Comment (ProntoPolice) [inline lib/app_context_config.rb]: Do not prefix writer method names with `set_`.

File: lib/parse_mobile_apps.rb (hunk 9 above)
Comment (ProntoPolice) [inline lib/parse_mobile_apps.rb]: Assignment Branch Condition size for from_csv_file is too high. [15.1/15]

Comment (ProntoPolice) [inline lib/parse_mobile_apps.rb]: Method has too many lines. [13/10]

File: lib/remove_context.rb (hunk 11 above)
Comment (ProntoPolice) [inline lib/remove_context.rb]: Method has too many lines. [12/10]

File: lib/remove_context.rb [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+  IO.foreach(filename) do |file_line|
+    domain = file_line.strip.downcase
+    iab_id = redis_connection.get(domain)
+    if iab_id
+      valid_entries << domain
+    else
+      invalid_entries << domain
+    end
+  end
+  return valid_entries, invalid_entries
```
Comment (ProntoPolice) [inline lib/remove_context.rb]: Redundant `return` detected.

File: lib/remove_context.rb [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    logger.info("#{domain} removed from database successfully.")
+  end
+  logger.info("Number of mobile app domains removed: #{valid_entries.length}")
+end
+
+def print_invalid_entries(invalid_entries, logger)
+  invalid_entries.each do |domain|
+    logger.warn("#{domain} not found thus NOT removed.")
+  end
+  logger.warn("Number of mobile app domains NOT removed: #{invalid_entries.length}")
```
Comment (ProntoPolice) [inline lib/remove_context.rb]: Line is too long. [84/80]

File: lib/remove_context.rb [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+end
+
+if __FILE__ == $PROGRAM_NAME
+  fail 'Usage: ruby remove_entries_from_file.rb config_file.yml' if ARGV.size != 1
```
Comment (ProntoPolice) [inline lib/remove_context.rb]: Always use `raise` to signal exceptions.

Comment (ProntoPolice) [inline lib/remove_context.rb]: Line is too long. [82/80]

File: spec/app_context_config_spec.rb [hunk 17]
```diff
@@ -0,0 +1,13 @@
+require 'spec_helper'
+require 'app_context_config'
+
+describe AppContextConfig do
+  context '.set_from_file' do
+    it 'raises an error when not supplied with a config file' do
+      expect {
```
Comment (ProntoPolice) [inline spec/app_context_config_spec.rb]: Avoid using `{...}` for multi-line blocks.

File: spec/app_context_config_spec.rb [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+        AppContextConfig.new('bad_file.yaml')
+      }.to raise_error(RuntimeError, 'You did not supply a config file.')
+    end
+  end
+end
+
```
Comment (ProntoPolice) [inline spec/app_context_config_spec.rb]: 1 trailing blank lines detected.

File: spec/app_context_spec.rb [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+describe AppContext do
+  before(:all) do
+    AppContextSpec.populate_iab_categories_table
+  end
+
+  after(:all) do
+    AppContextSpec.cleanup_iab_categories_table
+  end
+
+  let(:config_filename) { File.expand_path('../../config/app_context_test_config.yml', __FILE__) }
```
Comment (ProntoPolice) [inline spec/app_context_spec.rb]: Line is too long. [98/80]

File: spec/app_context_spec.rb [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+  let(:app_context_config) { AppContextConfig.new(config_filename) }
+  let(:app_context_expire_time) { app_context_config.app_context_expire_time }
+  let(:redis) { Redis.new(host: 'localhost', port: 6379) }
+  let(:iab_categories) { IABCategories.new(AppContextSpec.postgres_connection) }
+  let(:mobile_apps) { ParseMobileApps.from_csv_file(app_context_config.load_context_filename) }
```
Comment (ProntoPolice) [inline spec/app_context_spec.rb]: Line is too long. [95/80]

File: spec/app_context_spec.rb [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+  let(:mobile_apps_hash) { MobileApp.to_hash(mobile_apps) }
+  let(:logger) { Logger.new('/dev/null') }
+  let(:app_context) do
```
Comment (ProntoPolice) [inline spec/app_context_spec.rb]: Trailing whitespace detected.

File: spec/app_context_spec.rb [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+  let(:app_context_config) { AppContextConfig.new(config_filename) }
+  let(:app_context_expire_time) { app_context_config.app_context_expire_time }
+  let(:redis) { Redis.new(host: 'localhost', port: 6379) }
+  let(:iab_categories) { IABCategories.new(AppContextSpec.postgres_connection) }
+  let(:mobile_apps) { ParseMobileApps.from_csv_file(app_context_config.load_context_filename) }
+  let(:mobile_apps_hash) { MobileApp.to_hash(mobile_apps) }
+  let(:logger) { Logger.new('/dev/null') }
+  let(:app_context) do 
+    AppContext.new(redis,
+                   iab_categories)
```
Comment (ProntoPolice) [inline spec/app_context_spec.rb]: Trailing whitespace detected.

File: spec/app_context_spec.rb [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+    it 'contains the correct number of pairs' do
+      expect(subject.size).to eq(11)
+    end
+  end
+
+  context '.load' do
+    it 'loads an entry into Redis properly' do
+      expect(redis.get('com.abcdigital.abcnews')).to eq('197')
+    end
+
```
Comment (ProntoPolice) [inline spec/app_context_spec.rb]: Trailing whitespace detected.

File: spec/app_context_spec.rb [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+    end
+  end
+
+  context '.load' do
+    it 'loads an entry into Redis properly' do
+      expect(redis.get('com.abcdigital.abcnews')).to eq('197')
+    end
+    
+    it 'loads the correct number of key/value pairs' do
+      expect(redis.dbsize).to eq(11)
```
Comment (ProntoPolice) [inline spec/app_context_spec.rb]: Trailing whitespace detected.

File: spec/iab_categories_spec.rb [hunk 25]
```diff
@@ -0,0 +1,32 @@
+# spec/iab_categories_spec.rb
+
+require 'matching_mocks'
+require 'spec_helper'
+require 'iab_categories'
+
+
```
Comment (ProntoPolice) [inline spec/iab_categories_spec.rb]: Extra blank line detected.

File: lib/remove_context.rb (hunk 1 above)
Comment (kevint-simplifi) [inline lib/remove_context.rb]: :goat:

Comment (kevint-simplifi) [inline lib/remove_context.rb]: :shark:

**Pull request closed without merging**